ledger/complete/mtrie/trie/trie.go

```diff
@@ -170,20 +178,32 @@ func (mt *MTrie) read(res *[]*ledger.Payload, head *node.Node, paths []ledger.Pa
 //   * subtries that remain unchanged are from the parent trie instead of copied.
 // UNSAFE: method requires the following conditions to be satisfied:
 //   * keys are NOT duplicated
+//   * requires _all_ paths to have a length of mt.Height bits.
+// CAUTION: `updatedPaths` and `updatedPayloads` are permuted IN-PLACE for optimized processing.
 // TODO: move consistency checks from MForest to here, to make API safe and self-contained
 func NewTrieWithUpdatedRegisters(parentTrie *MTrie, updatedPaths []ledger.Path, updatedPayloads []ledger.Payload) (*MTrie, error) {
 	parentRoot := parentTrie.root
 	updatedRoot := parentTrie.update(parentRoot.Height(), parentRoot, updatedPaths, updatedPayloads, nil)
+	if parentRoot == updatedRoot {
+		return parentTrie, nil
+	}
```

Comment:
After a forest calls `Update`, the new returned trie will be compared with the existing tries in the forest (based on root hash) and won't be added if it already exists. 
I think the `if else` here will only save creating a new mtrie structure (which is small) in the very edge case where the entire trie didn't change, but we would have to do the `if else` check for all cases. I suggest to keep the code without the check.

Reply:
updated in https://github.com/onflow/flow-go/pull/562/commits/628dbf1df7dead82fdc9f0096c8c95a615708e4b